### PR TITLE
bsp: Introduce bsp manual for i.MX 93

### PR DESCRIPTION
**Pull request by cstoidner**
This PR introduces the BSP manual for the i.MX 93.

That is accomplished with the two steps below:

**Step 1:**
Copy files from `bsp/imx8/` and `bsp/imx8/imx8mp/` as templates to a new folder `bsp/imx93/`, and adapt them for purpose of i.MX 93. Whenever justifiable from "i.MX 93" work-scope, apply adaptions in a general manner (e.g. with reST substitutions), to encourage re-use across various i.MX platforms.

**Step 2:**
Move files that could be used across various i.MX platforms from `bsp/imx93/` to a new folder `bsp/imx-common`, and potentially apply further generalization.

--

After that PR was merged i.MX 8 platform can also make use of the "imx-common" files, can add further generalization as needed, and can drop their local duplicates from `imx8/*` folders.

--

That PR is a continuation of PR https://github.com/phytec/doc-bsp-yocto/pull/154
All commits of PR #154 are here contained, but squashed in the two existing commits. Also correction and a few improvements of wording and format are contained.

### CODE REVIEW

**Comment by cstoidner** [comment]
Please note:
That PR is currently in draft mode, as there are some adaptions not finally yet complete. Especially documents concerning "peripherals" are not yet finalized. That includes files:
* `source/bsp/peripherals/audio.rsti`
* `source/bsp/peripherals/emmc.rsti`
* `source/bsp/peripherals/rtc.rsti`
* `source/bsp/imx9/peripherals/*`

Also, some more files might be added to `bsp/imx-common`.

File: source/bsp/imx9/peripherals/pin-muxing.rsti
Comment (andrejpicej) [inline source/bsp/imx9/peripherals/pin-muxing.rsti:1]: Identical to the i.MX8 one, can probably be moved to a common place.

File: source/bsp/imx9/peripherals/pxp.rsti [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+into a single processing engine:
+
+* Scaling
+* Color Space Conversion (CSC)
+* Secondary Color Space Conversion (CSC2)
+* Rotation
+
+and thus minimizes the memory footprint required for the display pipeline.
+How to use the PXP with Gstreamer and Wayland check the `How to Use PXP in
+GStreamer and Wayland` (AN13829) Application note from NXP.
```
Comment (andrejpicej) [inline source/bsp/imx9/peripherals/pxp.rsti:14]: \ No newline at end of file. Again my fault. Sorry :smiling_face_with_tear:

File: source/bsp/imx9/peripherals/sd-card.rsti
Comment (andrejpicej) [inline source/bsp/imx9/peripherals/sd-card.rsti:1]: Can probably be moved to a common place, since the only difference is using the substitutions.

**Review by andrejpicej — CHANGES_REQUESTED** [review]
Thanks, looks better and more organized :). Some comments and suggestions though :).

File: source/bsp/imx9/peripherals/pm.rsti [hunk 2]
```diff
@@ -0,0 +1,140 @@
+Power Management
+----------------
+
+CPU Core Frequency Scaling
+..........................
+
+The CPU in the |soc| SoC is able to scale the clock frequency and the voltage.
+This is used to save power when the full performance of the CPU is not needed.
+Unlike i.MX8 M family the i.MX 93 doesn't support *Dynamic* Voltage and
```
Comment (ymoog) [inline source/bsp/imx9/peripherals/pm.rsti:9]: This file is in the imx9 directory, but you speak of the i.MX 93. It should probably be moved to the imx93 directory, then.

File: source/bsp/imx9/peripherals/canfd.rsti
Comment (ymoog) [inline source/bsp/imx9/peripherals/canfd.rsti:1]: Afaik the imx93 uses the same flexcan driver as imx8mp so we can probably make this file generic?

File: source/bsp/imx9/peripherals/display.rsti
Comment (ymoog) [inline source/bsp/imx9/peripherals/display.rsti:1]: Let's make this generic as well. We can restructure imx8 if necessary.

File: source/bsp/imx9/peripherals/eeprom.rsti [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+select the correct RAM configuration. This makes it possible to use the same
+bootloader image for different RAM sizes and choose the correct DTS overlays
+automatically.
+
+If the first 265 bytes of the normal area are deleted, the bootloader will fall
+back to the |som| Kit RAM setup, which is |kit-ram-size| RAM.
+
+.. warning::
+
+   Data in the first 265 bytes of the normal EEPROM area (bus: I2C-0 addr: 0x51)
```
Comment (ymoog) [inline source/bsp/imx9/peripherals/eeprom.rsti:56]:
```suggestion
   Data in the first 265 bytes of the normal EEPROM area (bus: I2C-2 addr: 0x50)
```

File: source/bsp/imx9/peripherals/eeprom.rsti [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+   shouldn't be erased or corrupted! This might influence the behavior of the
+   bootloader. The board might not boot correctly anymore.
+
+SoMs that are flashed with data format API revision 2 will print out information
+about the module in the early stage.
```
Comment (ymoog) [inline source/bsp/imx9/peripherals/eeprom.rsti:61]: I hope you do not have any SoM with EEPROM API < 2. I assume you can omit this paragraph?!

**Review by ymoog — CHANGES_REQUESTED** [review]
What about naming the imx directory ``imx`` instead of ``imx-common``. This would be consistent with other dir naming. e.g. we have a ``bsp`` dir not a ``bsp-common`` dir.

I agree with Andrej, we can do better with regards to creating common flies.

dt-overlay file for example could be structured in a way that first all the common operations that can be done with overlays are explained. And after that we explain the soc specific overlays (in a different soc specific file).
Maybe also for gpio, led and tm?

File: source/bsp/getting-started.rsti [hunk 5]
```diff
@@ -153,13 +156,25 @@ WIC image, so bmaptool knows which blocks to write and which to skip.
...
    :substitutions:
 
    host:~$ sudo dd if=|yocto-imagename|-|yocto-machinename|.wic of=/dev/<your_device> bs=1M conv=fsync status=progress
 
+Or to flash a compressed images (\*.wic.xz) use that command:
+
+.. code-block:: console
+   :substitutions:
+
+   host:~$ zstdcat |yocto-imagename|-|yocto-machinename|.wic.xz | sudo dd of=/dev/<your_device> bs=1M conv=fsync status=progress
```
[the remark this answers is not in the document]
Comment (cstoidner) [reply source/bsp/getting-started.rsti:176]: Sorry, I actually missed that discussion in the first PR. Sorry @BHahn42618  !
Let me address this here, I will link then that discussion from here into the first PR. 

Let me refer/response to the points from @BHahn42618  in the first PR (#154)

> in the documentation it says zstdcat is for compressing
> or uncompressing .zst files not .xz. Can this really be
> used for .xz files also?

That is indeed a good point. Looking into the man-pages one might assume the option `--format=xz` must be used to tell `zstdcat` that "xz"-algorithm shall be used. I used zstdcat for years and it works great, also for "xz" without any `--format` argument. Maybe it detects xz-algorithm due to the "xz" suffix of the file... However thats just an
assumption and might not work for all implementations of zstdcat (as it is not specified in the man-page). So thats a fair argument we need to consider.

> I would also expect a -d here for decompression, as it
> says default is compression.

Please be aware, we use `zstdcat`, not `zstd`. As you can see in the man-page `zstdcat` is queal `zstd -dcf`. That means it does always decompress (-d) and write to stdout (-c)

---

Apart from those details, I definitely perefer a method that pipes on the fly into `dd`, and avoids to create a physical file to uncompress the image. One advantage of that is you dont have to worry about a location to store that uncompressed file, that you will anyway delete immediately again after calling `dd`.

And especially when you use a `wic.xz` file from some public build location (e.g. a nightly build-pipeline, or an official release-folder, in our case `~/share/sw_data/Software/BSP-Yocto-*/**/`), the pipe-approach avoids hassles you might have:

- you have (hopefully) no write access to those public folders, so you have to think where to put the uncompressed file locally, and remove it again afterwards 
- or even critical: you have write-access to those public folder, then unxz's default behaviour is to remove the compressed xz-file and place the uncompressed version instead into that folder; 

To make it short, I feel the best solution is to use `xzcat` instead of `zstdcat` or `unxz`:
  - `xzcat` definitely uses the 'xz' algorithm
  - `xzcat` writes to stdout and so allows to pipe

That means:
  `host:~$ xzcat |yocto-imagename|-|yocto-machinename|.wic.xz | sudo dd of=/dev/<your_device> bs=1M conv=fsync status=progress`

[the remark this answers is not in the document]
Comment (cstoidner) [reply source/bsp/getting-started.rsti:176]: > We have a section in the manual about bmaptools. We should use that instead of dd for compressed/wic images.

@ymoog , we have not much changed here, compared to i.MX 8 manuals. Beside partup and bmaptools, `dd` is just the 3rd option to flash an image (thats the same in the i.MX 8 Manuals). Only difference here: Our `*.wic` is compressed, so it need to be uncompressed before giving to `dd`.

File: source/bsp/imx-common/development/format_sd-card.rsti
[the remark this answers is not in the document]
Comment (cstoidner) [reply source/bsp/imx-common/development/format_sd-card.rsti:1]: To keep it simple I would first like to put all equal things from imx8 and imx9 into the imx-common folder. Later we can decide to move individual things from imx-common even one level higher, if they are not imx specific.

I hope that's a good balance between speeding up the review/acceptance of that PR, by having also a good and future-proof structure.

File: source/bsp/imx-common/development/netboot.rsti
[the remark this answers is not in the document]
Comment (cstoidner) [reply source/bsp/imx-common/development/netboot.rsti:1]: Same as for the comment https://github.com/phytec/doc-bsp-yocto/pull/156#discussion_r1509312884 above.

File: source/bsp/imx-common/development/standalone_build.rsti
[the remark this answers is not in the document]
Comment (cstoidner) [reply source/bsp/imx-common/development/standalone_build.rsti:1]: I agree. So same as for the comment https://github.com/phytec/doc-bsp-yocto/pull/156#discussion_r1509312884 above.

File: source/bsp/imx-common/development/format_sd-card.rsti [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+      ...
+
+*  Unmount all SD Card partitions.
+*  Launch GParted:
+
+   .. code-block:: console
+
+      host:~$ sudo gparted
+
+.. image:: /bsp/imx8/images/gparted1.png
```
[the remark this answers is not in the document]
Comment (cstoidner) [reply source/bsp/imx-common/development/format_sd-card.rsti]: True. Solved now, see commit  d28d9adc20978f2d66b12fdd3d841985b1c190fb

NOTE: Before merging that PR, I will squash that commit into 0e779ccd795f9f68cefefc699fddb3a495b2f8e7

File: source/bsp/imx9/images/gparted1.png
[the remark this answers is not in the document]
Comment (cstoidner) [reply source/bsp/imx9/images/gparted1.png:1]: True. Solved now, see commit  291cb578c2aff8487e84dc29571547838d7cdf56

NOTE: Before merging that PR, I will squash that commit into 0e779ccd795f9f68cefefc699fddb3a495b2f8e7

File: source/bsp/imx-common/development/uuu.rsti [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+   UUU flashes U-boot into eMMC BOOT (hardware) boot partitions, and it sets
+   the BOOT_PARTITION_ENABLE in the eMMC! This is a problem since we want the
+   bootloader to reside in the eMMC USER partition. Flashing next U-Boot version
+   .wic image and not disabling BOOT_PARTITION_ENABLE bit will result in device
+   always using U-boot saved in BOOT partitions. To fix this in U-Boot:
+
+   .. code-block:: console
+
+      u-boot=> mmc partconf 0 0 0 0
```
[the remark this answers is not in the document]
Comment (cstoidner) [reply source/bsp/imx-common/development/uuu.rsti]: True. Solved now, see commit  e70d22b82c519c28fd4f4de2e3e6369ee07b6991

NOTE: Before merging that PR, I will squash that commit into f6bde8c6bf9ad168986059021f619c3c15336920

File: source/bsp/imx9/peripherals/eeprom.rsti (hunk 3 above)
Comment (cstoidner) [reply source/bsp/imx9/peripherals/eeprom.rsti]: Thanks for pointing out. Fixed in d1aa3a8f0a46e55ecdee6404781760eb3a251799

NOTE: Before merging that PR, I will squash that commit into f6bde8c6bf9ad168986059021f619c3c15336920

File: source/bsp/imx9/imx93/pd24.1.0.rst [hunk 8]
```diff
@@ -0,0 +1,448 @@
...
+
+On |sbc| the TI TLV320AIC3007 audio codec is used. It uses I2S for data
+transmission and I2C for codec control. The audio signals available are:
+
+* Stereo LINE IN,
+* Stereo LINE OUT,
+* Output where D-Class 1W speaker can be connected
+
+.. include:: /bsp/peripherals/audio.rsti
+	:end-before: .. audio_alsa_configuration_start_label
```
[the remark this answers is not in the document]
Comment (cstoidner) [reply source/bsp/imx9/imx93/pd24.1.0.rst:386]: Solved now, see commit  a24694f72fd562f8df3c2c35687d8dab4080f17d

NOTE: Before merging that PR, I will squash that commit into f6bde8c6bf9ad168986059021f619c3c15336920

File: source/bsp/getting-started.rsti (hunk 5 above)
[the remark this answers is not in the document]
Comment (BHahn42618) [reply source/bsp/getting-started.rsti:176]: @cstoidner I agree with you piping it directly into `dd` is a useful feature. Your suggestion with `xzcat` looks fine to me.

File: source/bsp/imx9/imx93/display.rsti [hunk 9]
```diff
@@ -0,0 +1,6 @@
+Display
+-------
+
+The |sbc| supports PEB-AV-02 with 7'' ``edt,etm0700g0edh6`` display with
+capacitive touchscreen. Overlay for said display is enabled in
+``BOOT/bootenv.txt`` by default!
```
[the remark this answers is not in the document]
Comment (cstoidner) [reply source/bsp/imx9/imx93/display.rsti]: Fixed now, see commit  9b70aaa2535c75f916f047ad81e3343c35e1fc63

NOTE: Before merging that PR, I will squash that commit into f6bde8c6bf9ad168986059021f619c3c15336920

File: source/bsp/imx9/peripherals/ocotp-ctrl.rsti [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+Reading Fuse Values in Linux
+............................
+
+To access the content of the fuses in Linux NXP provides the NVMEM_IMX_OCOTP
+module. All fuse content of the memory-mapped shadow registers is accessible via
+sysfs:
+
+.. code-block:: console
+
+   target:~$ hexdump /sys/devices/platform/soc\@0/47510000.efuse/fsb_s400_fuse0/nvmem
```
[the remark this answers is not in the document]
Comment (cstoidner) [reply source/bsp/imx9/peripherals/ocotp-ctrl.rsti]: Fixed now, see commit  9b70aaa2535c75f916f047ad81e3343c35e1fc63

NOTE: Before merging that PR, I will squash that commit into f6bde8c6bf9ad168986059021f619c3c15336920

File: source/bsp/imx9/peripherals/pxp.rsti (hunk 1 above)
Comment (cstoidner) [reply source/bsp/imx9/peripherals/pxp.rsti]: Fixed now, see commit  9b70aaa2535c75f916f047ad81e3343c35e1fc63

NOTE: Before merging that PR, I will squash that commit into f6bde8c6bf9ad168986059021f619c3c15336920

File: source/bsp/imx9/imx93/emmc.rsti
[the remark this answers is not in the document]
Comment (cstoidner) [reply source/bsp/imx9/imx93/emmc.rsti:1]: I agree. Done now with commit  2f010ce58f1ccfd09bfa7f0a8b601ddc06d4d03a

NOTE: Before merging that PR, I will squash that commit into 0e779ccd795f9f68cefefc699fddb3a495b2f8e7

File: source/bsp/imx9/peripherals/pin-muxing.rsti
Comment (cstoidner) [reply source/bsp/imx9/peripherals/pin-muxing.rsti:1]: I agree. Done now with commit  3bbf078b87ed5d7663dbf0f31d1b084abdd47c17

NOTE: Before merging that PR, I will squash that commit into 0e779ccd795f9f68cefefc699fddb3a495b2f8e7

File: source/bsp/imx9/peripherals/sd-card.rsti
Comment (cstoidner) [reply source/bsp/imx9/peripherals/sd-card.rsti:1]: I agree. Done now with commit  3bbf078b87ed5d7663dbf0f31d1b084abdd47c17

NOTE: Before merging that PR, I will squash that commit into 0e779ccd795f9f68cefefc699fddb3a495b2f8e7

File: source/bsp/imx9/peripherals/network.rsti
[the remark this answers is not in the document]
Comment (cstoidner) [reply source/bsp/imx9/peripherals/network.rsti:1]: I agree. Done with commit 77f2baaed5abf076718cc48ab16634119f551e6a

NOTE: Before merging that PR, I will squash that commit into 0e779ccd795f9f68cefefc699fddb3a495b2f8e7

File: source/bsp/imx9/imx93/pd24.1.0.rst [hunk 11]
```diff
@@ -0,0 +1,448 @@
...
+.. |emmcdev| replace:: mmcblk0
+.. |sdcarddev| replace:: mmcblk1
+
+
+.. Bootloader
+.. |u-boot-offset| replace:: 32
+.. |u-boot-offset-boot-part| replace:: 0
+.. |u-boot-mmc-flash-offset| replace:: 0x40
+
+.. IMX8(MP) specific
```
[the remark this answers is not in the document]
Comment (cstoidner) [reply source/bsp/imx9/imx93/pd24.1.0.rst]: Fixed, see 3580193e60915f1da16fea3f939a8e7d80668dcc

NOTE: Before merging that PR, I will squash that commit into f6bde8c6bf9ad168986059021f619c3c15336920

File: source/bsp/imx9/installing-os.rsti
[the remark this answers is not in the document]
Comment (cstoidner) [reply source/bsp/imx9/installing-os.rsti:1]: There are a few differences to `source/bsp/imx8/installing-os.rsti` that need to be considered to make it a common file. We should work together with i.MX 8 Team on it, when that i.MX 93 specific PR was done.

File: source/bsp/imx9/peripherals/canfd.rsti
Comment (cstoidner) [reply source/bsp/imx9/peripherals/canfd.rsti:1]: I dont think it is the same driver. The imx93 has flexcan, and imx8mp seems to use "mcp25xxfd". So lets go for now with a platform specific document. 

When that i.MX 93 specific PR was done, we both should check together if it makes sense establish here a common document.

File: source/bsp/imx9/peripherals/display.rsti
Comment (cstoidner) [reply source/bsp/imx9/peripherals/display.rsti:1]: Most of that document is speaking about backlight/brightness, which differs in imx8 and imx9. Lets go here for now with a platform specific document.

And again, when that i.MX 93 specific PR was done, we both should check together if it makes sense establish here a common document.

**Review by andrejpicej — APPROVED** [review]
Looks good to me. Like you wrote @cstoidner  there is still some generalization to be done. If i.MX8 team is OK with the current state this can then be done in separate PRs.

File: source/bsp/imx-common/development/netboot.rsti [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+   fdt_file=oftree
+   nfsroot=/srv/nfs
+   overlays=<overlayfilenames>
+
+<overlayfilenames> has to be replaced with the devicetree overlay filenames that you want to use.
+Separate the filenames by spaces. For example:
+
+.. code-block::
+
+   overlays=imx93-phyboard-segin-peb-av-02.dtbo imx93-phyboard-segin-peb-eval-01.dtbo
```
Comment (BHahn42618) [inline source/bsp/imx-common/development/netboot.rsti:184]: Not really common these names. Please use at least the already existing `|dt-carrierboard|` substitution instead of hardcoding `imx93-phyboard-segin`. Maybe also create a substitution for the `peb-av-02`. This would make it easier to include the file for the imx8 later on.

**Review by BHahn42618 — CHANGES_REQUESTED** [review]

File: source/bsp/imx-common/development/netboot.rsti (hunk 12 above)
Comment (cstoidner) [reply source/bsp/imx-common/development/netboot.rsti]: > Maybe also create a substitution for the peb-av-02. This would make it easier to include the file for the imx8 later on.

Not all platforms necessarily offer a `peb-av-02` overlay, as there is not standard set of DTBOs for all platforms. That makes it hard to follow your wish with substitutions here. Finally it is just a simple example that shows where/how to place the space characters between file-names.

I have now replaced both overlay file names with generic example file names `example-overlay1.dtbo` and `example-overlay2.dtbo`. I think that serves its purpose and avoids unnecessary complications.

See a15c8190ed8634712cd8610fea9cb3c31b65e18c

File: source/bsp/imx9/peripherals/eeprom.rsti (hunk 4 above)
Comment (cstoidner) [reply source/bsp/imx9/peripherals/eeprom.rsti]: Removed that sentence as we have only APIv2. See e4c3bdfc6ec5486e5d50f9a29406d88da5e74742

File: source/bsp/getting-started.rsti (hunk 5 above)
[the remark this answers is not in the document]
Comment (cstoidner) [reply source/bsp/getting-started.rsti]: Now `zstdcat` was replaced with `xzcat`, see 22f5386c40e826185d0ad4fd57c281e8d0e4b6e2

File: source/bsp/imx9/peripherals/pm.rsti (hunk 2 above)
Comment (cstoidner) [reply source/bsp/imx9/peripherals/pm.rsti]: File was now moved. See db8a0eae7c60d274b6485498c135a4950fcdc2f3

**Review by ymoog — APPROVED** [review]

**Review by pfiser — APPROVED** [review]

**Comment by jonas-rem** [comment]
@cstoidner is this PR ready to merge? If yes, can you click "Ready for review".

@BHahn42618 are you ok with the most recent changes? You had some remarks earlier.

If everything is good to go I will merge it later.

**Review by andrejpicej — APPROVED** [review]

**Comment by cstoidner** [comment]
> If everything is good to go I will merge it later.

It is now in "ready for review", and for me its even ready to be merged.
@jonas-rem , would you do a squash-mege? All the small fixing commits were useful for the review, but will be confusing in the commit-history of the official branch.
